sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SortOrder.scala

```diff
@@ -71,12 +95,33 @@ case class SortPrefix(child: SortOrder) extends UnaryExpression {
 
   val nullValue = child.child.dataType match {
     case BooleanType | DateType | TimestampType | _: IntegralType =>
-      Long.MinValue
+      if ((child.isAscending && child.nullOrdering == NullFirst) ||
+        (!child.isAscending && child.nullOrdering == NullLast)) {
+          Long.MinValue
+      } else {
+        Long.MaxValue
+      }
     case dt: DecimalType if dt.precision - dt.scale <= Decimal.MAX_LONG_DIGITS =>
-      Long.MinValue
+      if ((child.isAscending && child.nullOrdering == NullFirst) ||
+        (!child.isAscending && child.nullOrdering == NullLast)) {
+        Long.MinValue
+      } else {
+        Long.MaxValue
+      }
     case _: DecimalType =>
-      DoublePrefixComparator.computePrefix(Double.NegativeInfinity)
-    case _ => 0L
+      if ((child.isAscending && child.nullOrdering == NullFirst) ||
+        (!child.isAscending && child.nullOrdering == NullLast)) {
+        DoublePrefixComparator.computePrefix(Double.NegativeInfinity)
+      } else {
+        DoublePrefixComparator.computePrefix(Double.NaN)
```

Comment:
I tried the PositiveInfinity before but the randomized test sometimes fails me. Then, I view the actual sorted array and found that `NaN` is always considered bigger than `PositiveIntinity`. So I changed to use `NaN` here.